scalameta/semanticdb-scalac/src/main/scala/scala/meta/interactive/InteractiveSemanticdb.scala

```diff
@@ -0,0 +1,112 @@
+  def newCompiler(classpath: String, scalacOptions: List[String]): Global = {
+    val vd = new VirtualDirectory("(memory)", None)
+    val settings = new Settings
+    settings.outputDirs.setSingleOutput(vd)
+    settings.classpath.value = classpath
+    if (classpath.isEmpty) {
+      settings.usejavacp.value = true
+    }
+    settings.processArgumentString(
+      ("-Ypresentation-any-thread" :: scalacOptions).mkString(" ")
+    )
+    val compiler = new Global(settings, new StoreReporter)
+    compiler
+  }
+
+  def toDocument(compiler: Global, code: String): Document =
+    toDocument(compiler, code, "interactive.scala", 10000)
+
+  /**
+    * Build semanticdb document from this snippet of code.
+    *
+    * @param compiler an instance of scalac interactive global.
+    * @param code the code to be compiled.
+    * @param filename the name of the source file.
+    * @param timeout max number of milliseconds to allow the presentation compiler
+    *                to typecheck this file.
+    *  @throws Exception note that this method can fail in many different ways with excpet
+    */
+  def toDocument(compiler: Global, code: String, filename: String, timeout: Long): Document = {
+    val unit = addCompilationUnit(compiler, code, filename)
+    // reload seems to be necessary before askLoadedType.
+    ask[Unit](r => compiler.askReload(unit.source :: Nil, r)).get
+    val compiledTree =
+      ask[compiler.Tree](r => compiler.askLoadedTyped(unit.source, r))
+        .get(timeout)
+    val tree = compiledTree match {
+      case Some(Left(t)) => t
+      case Some(Right(ex)) => throw ex
+      case None => throw new IllegalArgumentException("Presentation compiler timed out")
+    }
+    lazy val databaseOps: DatabaseOps {
+      val global: compiler.type
+    } = new DatabaseOps {
+      val global: compiler.type = compiler
+    }
+    import databaseOps._
+    unit.body = tree
+    val document = unit.asInstanceOf[databaseOps.global.CompilationUnit].toDocument
+    document
+  }
+
+  /**
+    * Inserts "_CURSOR_" at given offset.
+    *
+    * _CURSOR_ hints to the presentation compiler that this file is being edited
+    * with the cursor at that offset. This hint helps completions amongst
+    * other things.
+    */
+  def addCursor(code: String, offset: Int): String = {
+    new StringBuilder(code.length + "_CURSOR_".length)
+      .append(code.substring(0, offset))
+      .append("_CURSOR_")
+      .append(code.substring(offset))
+      .toString()
+  }
+
+  /** Create new compilation unit from given code. */
+  def addCompilationUnit(
+      global: Global,
+      code: String,
+      filename: String
+  ): global.RichCompilationUnit = {
+    val unit = global.newCompilationUnit(code, filename)
+    val richUnit = new global.RichCompilationUnit(unit.source)
+    global.unitOfFile(richUnit.source.file) = richUnit
+    richUnit
+  }
+
+  private def thisClasspath: String = this.getClass.getClassLoader match {
```

Comment:
What's the environment in which this method is supposed to be used?

Reply:
whenever this classloader is URLClassloader like when testing in sbt? It's there for convenience to be able to quickly build semanticdbs without any ceremony.